### PR TITLE
Allow update of router nat

### PR DESCRIPTION
**Pull request by emilymye**
Fixes https://github.com/terraform-providers/terraform-provider-google/issues/4081
Fixes https://github.com/terraform-providers/terraform-provider-google/issues/2773

Helps with https://github.com/terraform-providers/terraform-provider-google/issues/2977

# Release Note for Downstream PRs (will be copied)
```releasenote
compute: Enable update for `google_compute_router_nat`
```

### CODE REVIEW

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 1]
```diff
@@ -78,80 +28,109 @@ func resourceComputeRouterNat() *schema.Resource {
...
+						},
+					},
+				},
 			},
-			"min_ports_per_vm": &schema.Schema{
+			"min_ports_per_vm": {
 				Type:     schema.TypeInt,
 				Optional: true,
-				ForceNew: true,
+				Computed: true,
```
Comment (emilymye) [inline third_party/terraform/resources/resource_compute_router_nat.go]: I don't think there is any reason to not have computed here - right now Cloud NAT doesn't return a value but hopefully they will? I think it's better to use Computed to reflect the value in the API if this isn't set in config

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 2]
```diff
@@ -78,80 +28,109 @@ func resourceComputeRouterNat() *schema.Resource {
...
 				Optional: true,
-				ForceNew: true,
 				Elem:     &schema.Schema{Type: schema.TypeString},
 			},
-			"source_subnetwork_ip_ranges_to_nat": &schema.Schema{
+			"source_subnetwork_ip_ranges_to_nat": {
 				Type:         schema.TypeString,
-				Optional:     true,
-				ForceNew:     true,
+				Required:     true,
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_compute_router_nat.go]: What happened if a user didn't set this before?

File: third_party/terraform/resources/resource_compute_router_nat.go (hunk 1 above)
Comment (rileykarson) [reply third_party/terraform/resources/resource_compute_router_nat.go]: I'm a little sad declaring a field O+C, but I don't think we have a better option so sgtm

Comment (rileykarson) [reply third_party/terraform/resources/resource_compute_router_nat.go]: What do you think of declaring a default in `3.X`? I bet the API uses a fixed value, even if it's a little vague

> Minimum number of ports allocated to a VM from this NAT config. If not set, a default number of ports is allocated to a VM. This is rounded up to the nearest power of 2. For example, if the value of this field is 50, at least 64 ports are allocated to a VM.

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 3]
```diff
@@ -78,80 +28,109 @@ func resourceComputeRouterNat() *schema.Resource {
...
 			},
-			"min_ports_per_vm": &schema.Schema{
+			"min_ports_per_vm": {
 				Type:     schema.TypeInt,
 				Optional: true,
-				ForceNew: true,
+				Computed: true,
 			},
-			"udp_idle_timeout_sec": &schema.Schema{
+			"udp_idle_timeout_sec": {
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_compute_router_nat.go]: I think this field and the other O+C ones below have well-defined defaults. I'm assuming the API returns `nil` if they take the default; does it do the same when explicit values are sent?

If we could `default_if_empty` these fields and set Terraform-side defaults, I'd feel a lot better than declaring more O+C fields.

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 4]
```diff
@@ -302,8 +279,105 @@ func resourceComputeRouterNatRead(d *schema.ResourceData, meta interface{}) erro
 	return nil
 }
 
-func resourceComputeRouterNatDelete(d *schema.ResourceData, meta interface{}) error {
+func resourceComputeRouterNatUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_compute_router_nat.go]: I like how the diff tool decided the empty line was the important thing to preserve here vs the function header lol

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 5]
```diff
@@ -302,8 +279,105 @@ func resourceComputeRouterNatRead(d *schema.ResourceData, meta interface{}) erro
...
+		Nats: nats,
+	}
+
+	log.Printf("[DEBUG] Updating router %s/%s with nats: %+v", region, routerName, nats)
+	op, err := config.clientComputeBeta.Routers.Patch(project, region, router.Name, obj).Do()
+	if err != nil {
+		return fmt.Errorf("Error patching router %s/%s: %s", region, routerName, err)
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s/%s/%s", project, region, routerName, natName))
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_compute_router_nat.go]:
```suggestion
```

Comment (rileykarson) [reply third_party/terraform/resources/resource_compute_router_nat.go]: This is just redundant, right?

File: third_party/terraform/tests/resource_compute_router_nat_test.go [hunk 6]
```diff
@@ -188,9 +223,107 @@ func testAccComputeRouterNatBasic(testId string) string {
 	`, testId, testId, testId, testId)
 }
 
+// Like basic but with extra resources
+func testAccComputeRouterNatBasicBeforeUpdate(randPrefix string) string {
+	return fmt.Sprintf(`
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_compute_router_nat_test.go]: nit: since we're using a preformatted string, we can left-align the HCL instead of indenting it as if it were Go code.

File: third_party/terraform/resources/resource_compute_router_nat.go (hunk 1 above)
Comment (emilymye) [reply third_party/terraform/resources/resource_compute_router_nat.go]: I'm ok with that. I also filed a request with compute

File: third_party/terraform/resources/resource_compute_router_nat.go (hunk 2 above)
Comment (emilymye) [reply third_party/terraform/resources/resource_compute_router_nat.go]: they would get an error from the API if it's not set, and we don't set a default - i think it's ok to treat it less like a breaking change and more like a bugfix?

File: third_party/terraform/resources/resource_compute_router_nat.go (hunk 4 above)
Comment (emilymye) [reply third_party/terraform/resources/resource_compute_router_nat.go]: 🤷‍♀

File: third_party/terraform/resources/resource_compute_router_nat.go (hunk 5 above)
Comment (emilymye) [reply third_party/terraform/resources/resource_compute_router_nat.go]: yeah, it was there before so I was like fewer changes I make the better, but I can remove it

File: third_party/terraform/resources/resource_compute_router_nat.go (hunk 3 above)
Comment (emilymye) [reply third_party/terraform/resources/resource_compute_router_nat.go]: the API JSON response doesn't have the fields but golang library means it returns zeros - your linked issue though suggested we shouldn't because its a breaking change to set defaults? so we might need to do it in v3

I'm also fine leaving it as just computed for now and making changes in 3.0.0 since I don't think it changes the behavior that much since rn the API doesn't return anything

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 7]
```diff
@@ -78,80 +28,105 @@ func resourceComputeRouterNat() *schema.Resource {
...
 				Type:     schema.TypeInt,
 				Optional: true,
-				ForceNew: true,
 			},
-			"tcp_established_idle_timeout_sec": &schema.Schema{
+			"tcp_established_idle_timeout_sec": {
 				Type:     schema.TypeInt,
 				Optional: true,
-				ForceNew: true,
+				Computed: true,
```
Comment (danawillow) [inline third_party/terraform/resources/resource_compute_router_nat.go]: Looks like this is the only `_timeout_sec` param that's marked Computed at the moment

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 8]
```diff
@@ -230,7 +204,6 @@ func resourceComputeRouterNatCreate(d *schema.ResourceData, meta interface{}) er
 	if err != nil {
 		return fmt.Errorf("Error patching router %s/%s: %s", region, routerName, err)
 	}
-	d.SetId(fmt.Sprintf("%s/%s/%s/%s", project, region, routerName, natName))
```
Comment (danawillow) [inline third_party/terraform/resources/resource_compute_router_nat.go]: why was this removed?

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 9]
```diff
@@ -302,8 +274,105 @@ func resourceComputeRouterNatRead(d *schema.ResourceData, meta interface{}) erro
...
+			return fmt.Errorf("Router %s not found", routerName)
+		}
+
+		return fmt.Errorf("Error reading parent router %s: %s", routerName, err)
+	}
+
+	var nat *computeBeta.RouterNat
+	nats := router.Nats
+	for _, nat = range nats {
+		if nat.Name == natName {
```
Comment (danawillow) [inline third_party/terraform/resources/resource_compute_router_nat.go]: what happens if it's not found?

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 10]
```diff
@@ -302,8 +274,105 @@ func resourceComputeRouterNatRead(d *schema.ResourceData, meta interface{}) erro
...
+	}
+
+	nat.MinPortsPerVm = int64(d.Get("min_ports_per_vm").(int))
+	nat.UdpIdleTimeoutSec = int64(d.Get("udp_idle_timeout_sec").(int))
+	nat.IcmpIdleTimeoutSec = int64(d.Get("icmp_idle_timeout_sec").(int))
+	nat.TcpEstablishedIdleTimeoutSec = int64(d.Get("tcp_established_idle_timeout_sec").(int))
+	nat.TcpTransitoryIdleTimeoutSec = int64(d.Get("tcp_transitory_idle_timeout_sec").(int))
+
+	if v, ok := d.GetOk("nat_ip_allocate_option"); ok {
+		nat.NatIpAllocateOption = v.(string)
```
Comment (danawillow) [inline third_party/terraform/resources/resource_compute_router_nat.go]: this field is required, so `ok` should always be true in here and then you can get rid of the if

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 11]
```diff
@@ -302,8 +274,105 @@ func resourceComputeRouterNatRead(d *schema.ResourceData, meta interface{}) erro
...
+	}
+
+	if v, ok := d.GetOk("subnetwork"); ok {
+		nat.Subnetworks = expandSubnetworks(v.(*schema.Set).List())
+	} else {
+		nat.Subnetworks = []*computeBeta.RouterNatSubnetworkToNat{}
+	}
+
+	if v, ok := d.GetOk("source_subnetwork_ip_ranges_to_nat"); ok {
+		nat.SourceSubnetworkIpRangesToNat = v.(string)
```
Comment (danawillow) [inline third_party/terraform/resources/resource_compute_router_nat.go]: same here

File: third_party/terraform/resources/resource_compute_router_nat.go [hunk 12]
```diff
@@ -302,8 +274,105 @@ func resourceComputeRouterNatRead(d *schema.ResourceData, meta interface{}) erro
...
+	}
+
+	if v, ok := d.GetOk("log_config"); ok {
```
Comment (danawillow) [inline third_party/terraform/resources/resource_compute_router_nat.go]: what do we want to do if the user removes `log_config` from their config?

File: third_party/terraform/resources/resource_compute_router_nat.go (hunk 2 above)
Comment (rileykarson) [reply third_party/terraform/resources/resource_compute_router_nat.go]: Yep! Thanks.

File: third_party/terraform/resources/resource_compute_router_nat.go (hunk 3 above)
Comment (rileykarson) [reply third_party/terraform/resources/resource_compute_router_nat.go]: Oh no that totally means explicit `0` and the default are indistinguishable doesn't it. Sounds good to me.
